src/mpid/ch4/src/ch4_coll_impl.h

```diff
@@ -594,14 +594,19 @@ MPL_STATIC_INLINE_PREFIX int MPIDI_Reduce_intra_composition_alpha(const void *se
         recvbuf = (void *) ((char *) recvbuf - true_lb);
     }
 
+    /* non-zero root needs to send from recvbuf if using MPI_IN_PLACE  */
+    intra_sendbuf = (sendbuf == MPI_IN_PLACE && root != 0) ? recvbuf : sendbuf;
```

Comment:
I found why the release_gather code was immune to the bug. It is due to https://github.com/pmodels/mpich/blob/8451885885e28bdec66173eab5ec30e6c6b2c08d/src/mpid/ch4/shm/posix/release_gather/nb_reduce_release_gather.h#L397-L399. Essentially it was doing the same recvbuf replacement inside the algorithm.

In general, are we allowed to have `sendbuf` and `recvbuf` overlap? Apparently all our reduce algorithms works with this replacement, but I wonder whether this is a potential hazard.